src/v_fmt.c

```diff
@@ -0,0 +1,481 @@
+    {
+        // write column offset to offset table
+        uint32_t offs = (uint32_t)(rover - output);
+        PUTLONG(col_offsets, offs);
+
+        // write column posts
+        for (int p = 0; p < array_size(columns[c].posts); p++)
+        {
+            // Write row offset
+            PUTBYTE(rover, columns[c].posts[p].row_off);
+
+            // Write number of pixels
+            int numpixels = array_size(columns[c].posts[p].pixels);
+            PUTBYTE(rover, numpixels);
+
+            // Write pad byte
+            byte lastval = numpixels ? columns[c].posts[p].pixels[0] : 0;
+            PUTBYTE(rover, lastval);
+
+            // Write pixels
+            for (int a = 0; a < numpixels; a++)
+            {
+                lastval = columns[c].posts[p].pixels[a];
+                PUTBYTE(rover, lastval);
+            }
+
+            // Write pad byte
+            PUTBYTE(rover, lastval);
+
+            array_free(columns[c].posts[p].pixels);
+        }
+
+        // Write 255 cap byte
+        PUTBYTE(rover, 0xff);
+
+        array_free(columns[c].posts);
+    }
+
+    array_free(columns);
+
+    // Done!
+    return (patch_t *)output;
+}
+
+static patch_t *DummyPatch(int lump, pu_tag tag)
+{
+    int num = (W_CheckNumForName)("TNT1A0", ns_sprites);
+    int len = W_LumpLength(num);
+    patch_t *dummy = W_CacheLumpNum(num, PU_CACHE);
+
+    Z_Malloc(len, tag, &lumpcache[lump]);
+    memcpy(lumpcache[lump], dummy, len);
+    return lumpcache[lump];
+}
+
+static void *DummyFlat(int lump, pu_tag tag)
+{
+    const int size = 64 * 64;
+
+    Z_Malloc(size, tag, &lumpcache[lump]);
+    memset(lumpcache[lump], v_darkest_color, size);
+    return lumpcache[lump];
+}
+
+// Set memory usage limits for storing standard and unknown chunks,
+// this is important when reading untrusted files!
+#define PNG_MEM_LIMIT (1024 * 1024 * 64)
+
+patch_t *V_CachePatchNum(int lump, pu_tag tag)
+{
+    if (lump >= numlumps)
+    {
+        I_Error("V_CachePatchNum: %d >= numlumps", lump);
+    }
+
+    if (lumpcache[lump])
+    {
+        Z_ChangeTag(lumpcache[lump], tag);
+        return lumpcache[lump];
+    }
+
+    void *buffer = W_CacheLumpNum(lump, tag);
+    int buffer_length = W_LumpLength(lump);
+
+    if (buffer_length < 8 || memcmp(buffer, "\211PNG\r\n\032\n", 8))
+    {
+        return buffer;
+    }
+
+    spng_ctx *ctx = spng_ctx_new(0);
+
+    // Ignore and don't calculate chunk CRC's
+    spng_set_crc_action(ctx, SPNG_CRC_USE, SPNG_CRC_USE);
+
+    spng_set_chunk_limits(ctx, PNG_MEM_LIMIT, PNG_MEM_LIMIT);
+
+    spng_set_option(ctx, SPNG_KEEP_UNKNOWN_CHUNKS, 1);
+
+    spng_set_png_buffer(ctx, buffer, buffer_length);
+
+    struct spng_ihdr ihdr = {0};
+    int ret = spng_get_ihdr(ctx, &ihdr);
+
+    if (ret)
+    {
+        I_Printf(VB_ERROR, "V_CachePatchNum: spng_get_ihdr %s\n",
+                 spng_strerror(ret));
+        goto error;
+    }
+
+    if (ihdr.color_type != SPNG_COLOR_TYPE_INDEXED || ihdr.bit_depth != 8)
```

Comment:
So, we assume that a paletted image is stored with the proper in-game palette.

Reply:
Yes. We can translate the palette with `I_GetPaletteIndex`, but I'm not sure it's needed.